gcp/warehouse.go

```diff
@@ -0,0 +1,160 @@
+	Project     string
+	DefaultRole string
+	Scopes      []string
+}
+
+// AccountWarehouse is used to create Google Cloud Platform Service Account
+// keys and access tokens associated with a specific identity.
+type AccountWarehouse struct {
+	opts  AccountWarehouseOptions
+	iam   *iam.Service
+	creds *iamcredentials.Service
+	crm   *cloudresourcemanager.Service
+}
+
+// NewAccountWarehouse creates a new AccountWarehouse using the provided client
+// and options.
+func NewAccountWarehouse(client *http.Client, opts *AccountWarehouseOptions) (*AccountWarehouse, error) {
+	iamSvc, err := iam.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM client: %v", err)
+	}
+
+	creds, err := iamcredentials.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating IAM credentials client: %v", err)
+	}
+
+	crm, err := cloudresourcemanager.New(client)
+	if err != nil {
+		return nil, fmt.Errorf("creating cloud resource manager client: %v", err)
+	}
+
+	return &AccountWarehouse{
+		opts:  *opts,
+		iam:   iamSvc,
+		creds: creds,
+		crm:   crm,
+	}, nil
+}
+
+// GetAccountKey returns a service account key associated with id.
+func (wh *AccountWarehouse) GetAccountKey(ctx context.Context, id string) ([]byte, error) {
+	account, err := wh.getBackingAccount(ctx, id)
+	if err != nil {
+		return nil, fmt.Errorf("getting backing account: %v", err)
+	}
+
+	// TODO(#2): Remove old keys from service account before creating new keys.
+	keys := wh.iam.Projects.ServiceAccounts.Keys
+	result, err := keys.Create(accountID("-", account), &iam.CreateServiceAccountKeyRequest{
+		PrivateKeyType: "TYPE_GOOGLE_CREDENTIALS_FILE",
+	}).Context(ctx).Do()
+	if err != nil {
+		return nil, fmt.Errorf("creating key: %v", err)
+	}
+
+	out, err := base64.StdEncoding.DecodeString(result.PrivateKeyData)
+	if err != nil {
+		return nil, fmt.Errorf("decoding key: %v", err)
+	}
+
+	return out, nil
+}
+
+// GetAccessToken returns an access token for the service account uniquely
+// associated with id.
+func (wh *AccountWarehouse) GetAccessToken(ctx context.Context, id string) (string, error) {
+	account, err := wh.getBackingAccount(ctx, id)
+	if err != nil {
+		return "", fmt.Errorf("getting backing account: %v", err)
+	}
+
+	response, err := wh.creds.Projects.ServiceAccounts.GenerateAccessToken(accountID("-", account), &iamcredentials.GenerateAccessTokenRequest{
+		Scope: wh.opts.Scopes,
+	}).Context(ctx).Do()
+	if err != nil {
+		return "", fmt.Errorf("generating access token: %v", err)
+	}
+
+	return response.AccessToken, nil
+}
+
+func (wh *AccountWarehouse) getBackingAccount(ctx context.Context, id string) (string, error) {
+	service := wh.iam.Projects.ServiceAccounts
+
+	hid := hashID(id)
+	name := accountID(wh.opts.Project, fmt.Sprintf("%s@%s.iam.gserviceaccount.com", hid, wh.opts.Project))
+	account, err := service.Get(name).Context(ctx).Do()
+	if err == nil {
+		return account.Email, nil
+	}
+	gErr, ok := err.(*googleapi.Error)
```

Comment:
why not
if err, ok := err.(*googleapi.Error); !ok || err.Code != http.StatusNotFound {
 ...
}